### ISSUE TITLE
Improve spec compliance of Blob Url store: resolve entry as part of URL parsing

### ISSUE THREAD

**Issue by gterzian** [servo/servo#25226]
We've got a couple of wpt failures that are the result of a "race condition" of sort between revoking and resolving a url. 

```
[url-with-fetch.any.worker.html]
  [Revoke blob URL after creating Request, will fetch]
    expected: FAIL


[url-with-fetch.any.html]
  [Revoke blob URL after creating Request, will fetch]
    expected: FAIL

[url-in-tags-revoke.window.html]
  expected: TIMEOUT
  [Fetching a blob URL immediately before revoking it works in an iframe.]
    expected: FAIL

  [Fetching a blob URL immediately before revoking it works in an iframe navigation.]
    expected: FAIL

  [Opening a blob URL in a new window immediately before revoking it works.]
    expected: TIMEOUT

  [Opening a blob URL in a noopener about:blank window immediately before revoking it works.]
    expected: FAIL

  [Opening a blob URL in a new window by clicking an <a> tag works immediately before revoking the URL.]
    expected: FAIL
```

And there a few new intermittents:

```
  ▶ Unexpected subtest result in /FileAPI/url/url-with-fetch.any.worker.html:
  │ FAIL [expected PASS] Revoke blob URL after calling fetch, fetch should succeed
  └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"

  ▶ Unexpected subtest result in /FileAPI/url/url-in-tags-revoke.window.html:
  │ TIMEOUT [expected PASS] Fetching a blob URL immediately before revoking it works in <script> tags.
  └   → Test timed out
```

What happens in all these cases is that the [resolving of the blob url](https://w3c.github.io/FileAPI/#blob-url-resolve) happens "too late", by the time the entry has already been [revoked](https://w3c.github.io/FileAPI/#removeTheEntry). 

"Resolving" the blob url is currently done inside [`scheme_fetch`](https://github.com/servo/servo/blob/bf587f22d7d9c96913a4df5added16f0454197f8/components/net/fetch/methods.rs#L611) whereas the URL spec says this should be done as part of URL parsing, at https://url.spec.whatwg.org/#ref-for-concept-url-blob-entry.

So that's why those WPT tests expect the fetch to succeed if for example the request is created before revoking the blob url, since per the spec by then the blob url should already have been resolved, even though fetch hasn't even started yet. 

The intermittency is due to the fact that a fetch is started on a thread-pool, while the revoking is done immediately when the `FileManagerThreadMsg::RevokeBlobURL` is handled. So in a given task, with the following operations:

1. Start a fetch with blob url,
2. Revoke blob url.

Then sometimes that fetch resolves the blob url before the revoke message is handled, and sometimes it doesn't.

I think this requires changing the resolve/revoke model of the `FileManagerStore`, where resolving means acquiring some sort of token, which would give you access to the underlying `FileImpl` even after is has been revoked. With revoking meaning you can't acquire anymore tokens.

**Comment by pshaughn** [comment]
I just noticed this issue now after having thought about this on a parallel track; #21517 and #13767 are both at least somewhat duplicates. My thoughts in #21517 are similar to the above, but more influenced by the way the spec treats the "parsed url" as an object in itself and has that object responsible for the blob reference.

**Comment by gterzian** [comment]
Yes I think you're right that the spec assumes that once you've parsed the url, you own the blob. In practice this "owning" will have to be implemented through some async cross-process workflow, for example as sketched above where you'd acquire a token representing ownership of the blob. 

Currently the cross-process workflow is broken in the sense that the blob url can be revoked even though it has been parsed(or is being parsed, and the parsing started before the revoking).

**Comment by pshaughn** [comment]
Do we need the actual Blob object on the other side of the handoff, or just the bytes from inside it and the MIME type? Would holding onto the latter with a reference count be easier than the former?

**Comment by gterzian** [comment]
Yes I think we only need the bytes, not the actual blob object. The object to keep alive is the `FileStoreEntry`. 

So there is already some reference counting going on, in response to `FileManagerThreadMsg::RevokeBlobURL` and `FileManagerThreadMsg::ActivateBlobURL`.

And the problem is that the blob url can be revoked before it has been resolved, in the code, not in the spec. The spec does this already when the url is parsed see https://url.spec.whatwg.org/#ref-for-concept-url-blob-entry

Our current implementation only does it in Fetch, at https://github.com/servo/servo/blob/bf587f22d7d9c96913a4df5added16f0454197f8/components/net/fetch/methods.rs#L764

So the tests that create a request(resulting in parsing the blob url), and then revoke the blob url, still expect the request to be successful(because it should already have been resolved). So we defenitely fail those, because we only do the resolving much later in fetch, and then the blob url has already been revoked. 

Then there are some intermittent tests, where the revoking is done not right after creating the request but after already having started it. There they can pass because the file manager will handle both operations in order, however the fetch is started on a separate thread, so the second message can be handled by the filemanager(and the entry removed), before the fetch in the other thread has had a chance to resolve the blob(which should actually have been done much before).

So the current reference counting works for the lifetime of the blob url(which should indeed be revoked), but it doesn't work for the lifetime of a fetch that is using the blob. We could just increase the count when the spec says to resolve the blob, maybe, but then we would also have to decrease either when the fetch is finished or aborted(or never started and the request garbage collected without having been used?). 

So basically the lifetime of the `FileStoreEntry` needs to be adjusted to better match the fact that the spec expects the blob to be available to a fetch once the blob url has been "resolved" for it, which actually happens very early.

**Comment by gterzian** [comment]
Note that I think it might make sense to first do https://github.com/servo/servo/issues/25723

Since tha could fix the intermittent test part of this one. It could also make it simpler to then "resolve the blob url when the url is parsed", since the `FileManager` would not be multithreaded like it is now.

**Comment by gterzian** [comment]
I'm trying to address this partially at https://github.com/servo/servo/pull/25724

**Comment by gterzian** [comment]
Ok I've made a first step at https://github.com/servo/servo/pull/25740

I see that there are a bunch of tests which will require acquiring this token much earlier, for example when a request is first created, and when a navigation starts. But that PR is a step in that direction, I've added a note at: https://github.com/servo/servo/pull/25740/commits/d08f888c8f8be0146a3ea5aa9271955dd31db9cc#diff-fb86627f1b3931ce6f2b706cd6603957R602

**Comment by jdm** [comment]
This may have an impact on #34320, since we see errors related to revoked blob URLs in the output there.

**Comment by jdm** [comment]
This also affects https://anubis.techaro.lol/, which is another turnstile-like intermediary.

**Comment by jdm** [comment]
I tried to implement a fix in https://github.com/servo/servo/compare/main...jdm:servo:blob-url-early-claim?expand=1 but ran into some difficulties. My idea was to extend ServoUrl to store an optional blob token that could automatically be revoked when the URL is dropped. This worked well enough to allow https://anubis.techaro.lol/ to load, but tests and cloudflare were still failing. After digging in, I realized that the design doesn't work—Arc-based correctness fails in the presence of IPC serialization.

What ends up happening is that my branch claims a blob token in the script thread when invoking an API like Request::Constructor. When the request is sent to the resource thread, however, we serialize the token rather than create another shared copy. This means that the URL in the script thread is dropped shortly afterwards and revokes the token, so the duplicate that exists in the resource thread is no longer valid.

Interesting sidebar: this is a case where using GenericChannel would probably allow this to work in single process mode, but multiprocess mode would still be broken.

**Comment by BergmannAtmet** [comment]
@jdm 

(I don't know anything)

Would something like using `PingDropTx`/`PingDropRx` wrappers not work? where you serialize a tx with the blob, and block on rx waiting for a ping drop from tx in the script thread in `PingDropRx::drop()`?

**Comment by jdm** [comment]
> [@jdm](https://github.com/jdm)
> 
> (I don't know anything)
> 
> Would something like using `PingDropTx`/`PingDropRx` wrappers not work? where you serialize a tx with the blob, and block on rx waiting for a ping drop from tx in the script thread in `PingDropRx::drop()`?

If I understand correctly, you're suggesting synchronizing the types across the channel boundary with another channel, to ensure that one doesn't get dropped too early. It's an interesting idea, but it doesn't work for this particular case because both sides should keep the blob alive and they are allowed to have different lifetimes.

I think an alternative solution here involves acquiring a new unique token before serializing a blob token, and ensuring that the new token value is immediately forgotten by the side that is performing the serialization. The side where it is deserialized will receive a valid token with a lifetime that is unrelated to the token that was used when sending.

**Comment by jdm** [comment]
> I think an alternative solution here involves acquiring a new unique token before serializing a blob token, and ensuring that the new token value is immediately forgotten by the side that is performing the serialization. The side where it is deserialized will receive a valid token with a lifetime that is unrelated to the token that was used when sending.

This appears to work; relevant WPT tests now pass and the blob-related errors no longer appear on the cloudfront turnstile. The implementation is definitely not intuitive, but at least we have a proof of existence. I need to decide how to clean up the branch (it's extremely janky and unreadable right now), or if there's another implementation possible that would be easier to understand/maintain.

Code: https://github.com/servo/servo/compare/main...jdm:servo:blob-url-early-claim?expand=1
Try run: https://github.com/servo/servo/actions/runs/17904848217

**Comment by simonwuelker** [comment]
> This appears to work; relevant WPT tests now pass and the blob-related errors no longer appear on the cloudfront turnstile. The implementation is definitely not intuitive, but at least we have a proof of existence. I need to decide how to clean up the branch (it's extremely janky and unreadable right now), or if there's another implementation possible that would be easier to understand/maintain.
> 
> Code: https://github.com/servo/servo/compare/main...jdm:servo:blob-url-early-claim?expand=1 Try run: https://github.com/servo/servo/actions/runs/17904848217

Do you mind if I try and get your branch in a state where we can merge it?

**Comment by jdm** [comment]
> > This appears to work; relevant WPT tests now pass and the blob-related errors no longer appear on the cloudfront turnstile. The implementation is definitely not intuitive, but at least we have a proof of existence. I need to decide how to clean up the branch (it's extremely janky and unreadable right now), or if there's another implementation possible that would be easier to understand/maintain.
> > Code: https://github.com/servo/servo/compare/main...jdm:servo:blob-url-early-claim?expand=1 Try run: https://github.com/servo/servo/actions/runs/17904848217
> 
> Do you mind if I try and get your branch in a state where we can merge it?

Please do!

**Comment by simonwuelker** [comment]
Hmm, one complication is that we now use `GenericChannels` in `resource_thread.rs` and those force the `ServoUrl` to send `CoreResourceMsg`s to the filemanager (instead of a untyped `IpcMessage`). But that can't happen because `net_traits` already depends on `servo_url`, causing a circular dependency.

 https://github.com/servo/servo/blob/099b8e0b80f3b9c0e33b9e5aa88c155a5200428a/components/net/resource_thread.rs#L265-L274

I wonder if `ServoUrl` could be generic over the token that is used, and then `net_traits` could re-export a variant of `ServoUrl` that knows about `CoreResourceMsg`... I've tried that but the diff quickly becomes enormous.

But probably trait objects are the way to go.
